### ISSUE TITLE
Very High Memory Usage of Grok Exporter.

### ISSUE THREAD

**Issue by Yagyansh** [fstab/grok_exporter#123]
Hi. I am using grok_exporter v 1.0.0.RC1.

global:
    config_version: 2
input:
    type: file
    path: access.log-20200807
    readall: true
grok:
    patterns_dir: grok_exporter-1.0.0.RC1.linux-amd64/patterns
metrics:
    - type: counter
      name: request_urls
      help: Total number of rejected recipients, partitioned by error message.
      match: '%{NGINX}'
      labels:
          method: '{{.request_type}}'
          path: '{{.pattern_url}}'
          code: '{{.response_code}}'
server:
    host: localhost
    port: 9144

My NGINX group pattern is below.
NGINX %{NUMBER:response_code} %{NUMBER:notrequired} %{NUMBER:notrequired} %{NUMBER:notrequired} \[%{HTTPDATE:access_time}\] \"(?:%{WORD:request_type} %{NOTSPACE:pattern_url}(?: HTTP/%{NUMBER:httpversion})?|-)\"

Grok Exporter is using constantly using 76% memory. Is it expected? Or am I doing something terribly wrong?

![image](https://user-images.githubusercontent.com/15364964/89646481-7e7f5b80-d8d9-11ea-890a-5379e83e9981.png)
PS - Iris mode is off, so these are absolute usage percentages.

![image](https://user-images.githubusercontent.com/15364964/89646540-9951d000-d8d9-11ea-82ca-cf264fb0dd4f.png)

**Comment by fstab** [comment]
grok_exporter should not need that much memory. Could you send me a few example log lines so I can reproduce this? Moreover, as you set `readall: true` grok_exporter will read the entire file on startup. This shouldn't be a problem, but to reproduce it, please let me know how many log lines are in this file.

**Comment by Yagyansh** [comment]
I kept readall: true because I did not take a live log file, I mean I took a chunk of logs and was trying grok exporter on it, so there was no live ingestion of logs in that file while the grok exporter was running. 

Number of log lines - 21052530
Example Logs - 

200 108 0.008 0.008 [07/Aug/2020:00:30:01 +0530] "GET /pattern1/pattern2/idblahblah HTTP/1.1" somedomainname - 47.31.35.72 2 10.10.1.88 - - - "-" 20940029 27 "upstream_cache_status=-" "-" "-" "Dalvik/2.1.0 (Linux; U; Android 6.0; A72 Build/MRA58K)" "47.31.35.72, 184.85.220.154, 96.17.150.151" "no-cache, max-age=0" "dhDebug=appVersion=1.0.8&osVersion=6.0&clientId=c1578f6a-5195&androidId=4c2a8c0bf3795104&installSource=sourcename&user_lang=en&nav_lang=en&width=720&height=1280" "1.1 v1-text(text) (Text), 1.1 Text.text(text) (Text)" - -userId- install_referrer= 

404 209 0.004 0.004 [06/Aug/2020:07:52:30 +0530] "GET /somepattern HTTP/1.1" 10.10.2.87 - - - 10.10.2.38 - - - "Text" 383 3 "upstream_cache_status=-" "-" "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36" "-" "-" "dhDebug=-" "-" - -userId-- install_referrer=-

**Comment by Yagyansh** [comment]
Also, what is the general Memory and CPU Utilization of Grok Exporter? I know it will depend on the grok pattern being used and the size of log file, but still, is there any benchmarking?

**Comment by fstab** [comment]
`grok_exporter` is basically a tool for executing regular expressions (grok patterns are converted to regular expressions internally). Almost all of the CPU usage should come from evaluating the regular expressions. There shouldn't be any significant overhead by grok_exporter itself. Memory usage should be minimal: The only time memory is used is if log lines are written faster than they are processed, because then grok_exporter buffers them in memory. The assumption here is that log lines are processed faster than new log lines are written, so there can be only temporary situations where this buffer is needed.

That said, I feel that grok_exporter should do a better job making it clear to the user where CPU and memory is spent. I started implementing some additional performance monitoring metrics that should make it obvious what CPU and memory are used for. I'm going to comment here as soon as this is available, and I hope this will help to see why memory usage is high in your case.

**Comment by Yagyansh** [comment]
Thanks a lot, that helps.

But one query here, as I mentioned before, my log file is static, no new logs are being ingested in my log file i.e no new log lines are being added at all. So, why am I getting warnings that log lines are written faster than they are processed, in turn spiking up the memory consumption?

**Comment by fstab** [comment]
Good point. Yes, with `readall: true` all initial lines will be in the buffer until they are processed.

I could implement a limit to the buffer size. I didn't do that yet because in real logging scenarios if log lines are written at a higher rate than they are processed, you are lost anyway, and I figured letting grok_exporter run out of memory is the best thing I can do. However, in your case a buffer size limit might actually make sense.

**Comment by Yagyansh** [comment]
Got it. Is implementing the feature of limiting buffer size in your pipeline in the near future?
Also, does this mean that once all the log files are read, my memory consumption should come down to normal?

**Comment by fstab** [comment]
> does this mean that once all the log lines are read, my memory consumption should come down to normal?

yes
> is implementing the feature of limiting buffer size in your pipeline in the near future?

I cannot promise that. There is already an (undocumented) config option `max_lines_in_buffer` that can be used to limit the buffer size in scenarios where log lines are written at a higher frequency than they can be processed. However, the implementation will drop the buffer and skip lines when the limit is reached. This is useful in some scenarios, but probably not in yours. Limiting the buffer and still guaranteeing that all log lines are processed means that the file tailer will fall behind reading the file, which might be a problem if logrotate trunkates the file while not all lines are read. I need to think this through.

Anyway, there might be a workaround with a bit of shell scripting: You could split the large file into a couple of smaller files and write a shell script that copies one small file after another into the log directory, waiting until grok_exporter has processed each file before copying the next one. In order to wait until grok_exporter has processed one file, you could `curl http://xps:9144/metrics | grep grok_exporter_lines_total` and check if the number of lines has increased by the expected number of lines in one small file. This should decrease memory usage as grok_exporter will only read one small file at a time.

**Comment by safebear** [comment]
I've noticed a grok exporter has a memory leak on our systems.  I'll post more info when I get a chance.
